src/electionguard/scheduler.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import annotations
+from contextlib import AbstractContextManager
+from multiprocessing import Pool as ProcessPool
+from multiprocessing.dummy import Pool as ThreadPool
+from multiprocessing.pool import Pool
+from psutil import cpu_count
+import traceback
+from typing import Any, Callable, Iterable, List
+
+from .logs import log_warning
+from .singleton import Singleton
+from .utils import T
+
+
+class Scheduler(Singleton, AbstractContextManager):
+    """
+    Worker that wraps Multprocessing and allows 
+    for shared context or spawning processes.
+    Implemented as a singleton to guarantee there is only one set
+    of tread and process pools in use throughout the library.
+    Also implements the [Context Manager Protocol](https://docs.python.org/3.8/library/stdtypes.html#typecontextmanager)
+    """
+
+    __process_pool: Pool
+    __thread_pool: Pool
+
+    def __init__(self) -> None:
+        super(Scheduler, self).__init__()
+        self._open()
+
+    def __enter__(self) -> Scheduler:
+        self._open()
+        return self
+
+    def __exit__(self, exc_type: Any, exc_value: Any, exc_traceback: Any) -> None:
+        self.close()
+
+    def _open(self) -> None:
+        """Open pools"""
+        max_processes = cpu_count(logical=False)
+        # Reserve one CPU for I/O bound tasks
+        if max_processes > 2:
+            max_processes = max_processes - 1
+        self.__process_pool = ProcessPool(max_processes)
+        self.__thread_pool = ThreadPool(max_processes)
+
+    def close(self) -> None:
+        """Close pools"""
+        self.__process_pool.close()
+        self.__thread_pool.close()
+
+    def cpu_count(self) -> int:
+        """Get CPU count"""
+        return cpu_count(logical=False)
```

Comment:
using this version of cpu_count allows us to specify only using physical cores.  Adding logical cores doesn't really get us anything for the work we are trying to schedule on most CPU's - and in fact the overhead can actually decrease performance depending on the CPU architecture and what data is flowing into the caches